File: ivay-frontend/src/service/product.service.js
import axios from "axios";
import { API_BASE_URL } from "../constants/api";

/**
 * Base URL for product related API endpoints.
 */
const PRODUCT_API_URL = `${API_BASE_URL}/products`;

/**
 * Fetches all products from the API.
 * @returns {Promise<Array<Object>>} Array of product response DTOs.
 */
const getAllProducts = async () => {
  const response = await axios.get(PRODUCT_API_URL);
  return response.data.data;
};

/**
 * Finds products by name using a filter endpoint.
 * @param {string} name - The name or part of the name to search for.
 * @returns {Promise<Array<Object>>} Array of product response DTOs matching the name.
 */
const findProductsByName = async (name) => {
  const response = await axios.get(`${PRODUCT_API_URL}/filter`, {
    params: { name },
  });
  return response.data.data;
};

/**
 * Fetches a specific product by its ID.
 * @param {number|string} productId - The ID of the product.
 * @returns {Promise<Object>} Product response DTO.
 */
const getProductById = async (productId) => {
  const response = await axios.get(`${PRODUCT_API_URL}/${productId}`);
  return response.data.data;
};

/**
 * Creates a new product.
 * @param {Object} productRequestDto - The data required to create a product.
 * @returns {Promise<Object>} The created product response DTO.
 */
const createProduct = async (productRequestDto) => {
  const response = await axios.post(PRODUCT_API_URL, productRequestDto);
  return response.data.data;
};

/**
 * Updates an existing product by its ID.
 * @param {number|string} productId - The ID of the product to update.
 * @param {Object} productRequestDto - The updated product data. Same structure as createProductRequestDto.
 * @returns {Promise<Object>} The updated product response DTO.
 */
const updateProduct = async (productId, productRequestDto) => {
  const response = await axios.put(
    `${PRODUCT_API_URL}/${productId}`,
    productRequestDto
  );
  return response.data.data;
};

/**
 * Deletes a product by its ID.
 * @param {number|string} productId - The ID of the product to delete.
 * @returns {Promise<void>}
 */
const deleteProduct = async (productId) => {
  await axios.delete(`${PRODUCT_API_URL}/${productId}`);
};

/**
 * Fetches all order items associated with a specific product.
 * @param {number|string} productId - The ID of the product.
 * @returns {Promise<Array<Object>>} Array of order item response DTOs.
 */
const getOrderItemsByProductId = async (productId) => {
  const response = await axios.get(
    `${PRODUCT_API_URL}/${productId}/order-items`
  );
  return response.data.data;
};

/**
 * Fetches all cart items associated with a specific product.
 * @param {number|string} productId - The ID of the product.
 * @returns {Promise<Array<Object>>} Array of cart item response DTOs.
 */
const getCartItemsByProductId = async (productId) => {
  const response = await axios.get(
    `${PRODUCT_API_URL}/${productId}/cart-items`
  );
  return response.data.data;
};

/**
 * Fetches products with pagination.
 * @param {number} [page=0] - The page number (0-indexed).
 * @param {number} [size=20] - The number of items per page.
 * @returns {Promise<Object>} A pagination object containing an array of product response DTOs and pagination info (like total elements, total pages, etc.).
 */
const getProductsPaginated = async (page = 0, size = 20) => {
  const response = await axios.get(`${PRODUCT_API_URL}/paginated`, {
    params: { page, size },
  });
  return response.data.data;
};

/**
 * Service object for interacting with Product related API endpoints.
 */
const ProductService = {
  getAllProducts,
  findProductsByName,
  getProductById,
  createProduct,
  updateProduct,
  deleteProduct,
  getOrderItemsByProductId,
  getCartItemsByProductId,
  getProductsPaginated,
};

export default ProductService;

File: ivay-backend/src/main/java/com/ivay/service/ProductService.java
package com.ivay.service;

import java.util.List;

import com.ivay.dtos.cartitemdto.CartItemResponseDto;
import com.ivay.dtos.orderitemdto.OrderItemResponseDto;
import com.ivay.dtos.productdto.PaginatedProductResponseDto;
import com.ivay.dtos.productdto.ProductRequestDto;
import com.ivay.dtos.productdto.ProductResponseDto;

/**
 * Service interface for managing products.
 *
 * Provides methods to:
 * - retrieve all products or a paginated subset
 * - search products by name
 * - perform CRUD operations on products
 * - retrieve associated order items and cart items for a product
 *
 * @since 1.0.0
 */
public interface ProductService {

    /**
     * Retrieves all products in the system.
     *
     * @return a list of {@link ProductResponseDto} representing all products
     */
    List<ProductResponseDto> getAllProducts();

    /**
     * Retrieves a page of products with pagination metadata.
     *
     * @param page zero-based page index
     * @param size number of items per page
     * @return a {@link PaginatedProductResponseDto} containing the page of products and metadata
     */
    PaginatedProductResponseDto getProductsPaginated(int page, int size);

    /**
     * Finds products whose name contains the given substring, case-insensitive.
     *
     * @param name substring to search for within product names
     * @return a list of matching {@link ProductResponseDto}
     */
    List<ProductResponseDto> findProductsByName(String name);

    /**
     * Retrieves a single product by its identifier.
     *
     * @param productId the identifier of the product to retrieve
     * @return the {@link ProductResponseDto} for the given product
     */
    ProductResponseDto getProductById(Long productId);

    /**
     * Creates a new product.
     *
     * @param productRequestDto the data for the new product
     * @return the created {@link ProductResponseDto}
     */
    ProductResponseDto createProduct(ProductRequestDto productRequestDto);

    /**
     * Updates an existing product.
     *
     * @param productId the identifier of the product to update
     * @param productRequestDto the new data for the product
     * @return the updated {@link ProductResponseDto}
     */
    ProductResponseDto updateProduct(Long productId, ProductRequestDto productRequestDto);

    /**
     * Deletes a product by its identifier.
     *
     * @param productId the identifier of the product to delete
     */
    void deleteProduct(Long productId);

    /**
     * Retrieves all order items that include a specific product.
     *
     * @param productId the identifier of the product
     * @return a list of {@link OrderItemResponseDto} containing that product
     */
    List<OrderItemResponseDto> getOrderItemsByProductId(Long productId);

    /**
     * Retrieves all cart items that include a specific product.
     *
     * @param productId the identifier of the product
     * @return a list of {@link CartItemResponseDto} containing that product
     */
    List<CartItemResponseDto> getCartItemsByProductId(Long productId);
}
